Add concepts for ontological analysis and unpacking with foundational ontology variants

pedropaulofb opened the pull request:
# add concepts for ontological analysis and unpacking with foundational ontology variants

**Description:**
This PR introduces a new set of concepts into the *Semantic Mapping Vocabulary* that capture the role of **ontological analysis** and **ontological unpacking**, together with their variants grounded in foundational ontologies (UFO, DOLCE, GFO, BFO).

## Motivation

The use of **ontological foundations** has grown substantially across diverse areas such as conceptual modeling, semantic interoperability, biomedical informatics, etc. These practices rely on foundational ontologies to make explicit the real-world semantics of information artifacts, improving rigor, consistency, and cross-domain comparability.

By including these concepts in SEMAPV, the vocabulary better supports mapping and analysis approaches that are increasingly adopted in ontology-driven and model-based research and practice.

## Relation to Existing Work

- **OntoUML/UFO Catalog + gUFO**:
  The [OntoUML/UFO catalog](https://w3id.org/ontouml-models/git) is widely used in the ontology community as an empirical source of models that apply ontological foundations (notably UFO), all of which can be represented in Semantic Web format through [gUFO](https://nemo-ufes.github.io/gufo/).
  - The proposed terms directly align with the analysis techniques applied in this catalog, enabling ontologically well-founded mappings between models and supporting their use as anchors for defining the real-world semantics of terms.

- [**Health-RI Semantic Interoperability Initiative**](https://w3id.org/health-ri/semantic-interoperability):
  Foundational ontology–based analysis strengthens the initiative’s **mapping strategy**, which emphasizes creating mappings to achieve semantic interoperability.
  - The new SEMAPV terms provide a formal vocabulary for describing these ontology-driven mapping techniques, thereby supporting consistent, FAIR, and reusable mappings across biomedical data sources.

- [**Open Biomedical and Biomedical Ontologies (OBO) Foundry**](https://obofoundry.org/):
  A large collection of domain ontologies in life sciences that can benefit from the new mapping techniques.
  - By providing precise terms for creating mapping to OBO terms, the SEMAPV extensions allow for **more precise mappings**.

## Summary of Changes

- Added `semapv:OntologicalAnalysis` and `semapv:OntologicalUnpacking` concepts.
- Added foundational variants corresponding to the **four most widely used foundational ontologies**: UFO-based, DOLCE-based, GFO-based, and BFO-based.
- Concepts include preferred and alternative labels, definitions, reference ([Guizzardi & Guarino, 2024](https://doi.org/10.1016/j.datak.2024.102325)), and illustrative examples.

## Impact

- Extends SEMAPV to cover the emerging demand for mappings built on ontology-based analysis techniques.
- Provides a semantic anchor for referencing foundational ontology practices in conceptual modeling and semantic integration.
- Prepares SEMAPV for broader adoption in communities where ontological foundations are a prerequisite for trustworthy mappings and analyses.

### Note to Maintainers

I did not run the Makefile to regenerate the OWL and HTML documentation files.
Could you kindly run the build process after this PR so that the generated artifacts reflect the new concepts?

I would like to thank the maintainers in advance for considering this extension.

I am happy to adjust the proposal or discuss any changes needed so that these additions align well with the goals of SEMAPV.

pedropaulofb committed:
feat(terms): add concepts for ontological analysis and unpacking

Added new SKOS concepts to `semapv-terms.tsv` covering Ontological Analysis and Ontological Unpacking, including UFO, DOLCE, GFO, and BFO based variants. Includes labels, definitions, sources, and examples.

These additions extend the vocabulary with foundational ontology–based approaches, aligned with the SEMAPV structure.

pedropaulofb committed:
chore(terms): fix missing newline at end of semapv-terms.tsv file

cthoyt commented:
hi @pedropaulofb, I am not really sure what these terms are or if they would fit into semapv based on what you've written.

I don't think the phrases you're using, namely `ontological analysis` and `ontological unpacking`, are common phrases. Could you please explain what they mean?

Is the goal to use the terms you've suggested in mapping justifications in SSSOM files? Can you give concrete, self-contained, explicit examples of where you might use the classes you are proposing? I.e., demonstration of their use in some kind of machine-readable artifacts.

I would also suggest cutting your PR down to a minimum number of classes for a first pass. Having completion over a variety of different upper level ontologies in an initial PR makes a lot of noise. Maybe just stick to how this looks in one or two, like BFO. If we move forward with it, it can be extended/followed up later

pedropaulofb commented:
Dear @cthoyt, thank you for the quick response and for the feedback.

I would like to add some background on the motivation for this proposal. The notion of *ontological analysis* has been present in the Information Sciences community since the 1990s, and it gained significant visibility with Guarino’s work, particularly in [1] (2,175 citations) and [2] (1,595 citations), and with the proceedings of the first specialized conference on formal ontology in information systems (FOIS 1998 [3]) (7,153 citations). A quick search on Google Scholar for the term "ontological analysis" returns about [20,800 results](https://scholar.google.com/scholar?q=%22ontological+analysis%22), which shows its relevance.

Since the early 2000s, much of the attention in our field has shifted toward the Semantic Web. However, even today there is no straightforward way (as far as I know) to indicate that a given mapping was created using ontological analysis. This is precisely the gap I am aiming to address with the current proposal.

Regarding your comment about specific types, in theory the proposed *Ontological Analysis* term without reference to a specific foundational ontology would be abstract, since there is no way to perform an ontological analysis without applying a foundational ontology as a grounding reference. I included it (the generic term) because there are several foundational ontologies in use that I did not explicitly accommodate with subtypes (e.g., SUMO, YAMATO, etc.). In the Semantic Web community, *foundational ontologies* are also called *upper-level ontologies*.

As a reviewer in some relevant academic conferences, I frequently encounter papers applying ontological analysis with the three most widely used foundational ontologies: UFO, BFO, and DOLCE. To give a sense of their prevalence, I did a quick Google Scholar search using the following queries:

| Search term | Results |
|-------------|-----------------|
| (UFO OR OntoUML OR gUFO) AND "ontological analysis" | [About 2,390](https://scholar.google.com/scholar?q=(UFO+OR+OntoUML+OR+gUFO)+AND+%22ontological+analysis%22) |
| (BFO OR "basic formal ontology") AND "ontological analysis" | [About 790](https://scholar.google.com/scholar?q=(BFO+OR+%22basic+formal+ontology%22)+AND+%22ontological+analysis%22) |
| (dolce OR "Descriptive Ontology for Linguistic") AND "ontological analysis" | [About 1,120](https://scholar.google.com/scholar?q=(dolce+OR+%22Descriptive+Ontology+for+Linguistic%22)+AND+%22ontological+analysis%22) |
| (GFO OR "General formal ontology") AND "ontological analysis" | [About 312](https://scholar.google.com/scholar?q=(GFO+OR+%22General+formal+ontology%22)+AND+%22ontological+analysis%22) |

Given these numbers, it might make sense to remove GFO, as its usage is significantly lower. Still, in essence, it would be better to explicitly support UFO, BFO, and DOLCE, since people performing mappings will almost always rely on one of these rather than using the generic class. After all, the purpose is to define precise semantics, so why would someone not specify the foundational ontology they are using?

I appreciate your consideration, as well as that of the other curators, in reviewing these additional points. I am, of course, open to further discussion and happy to adjust the proposal if needed.

**References**
[[1]](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=3-WRYO4AAAAJ&cstart=100&pagesize=100&sortby=pubdate&citation_for_view=3-WRYO4AAAAJ:u-x6o8ySG0sC) Guarino, N., & Giaretta, P. (1995). Ontologies and knowledge bases. Towards very large knowledge bases, 1-2.
[[2]](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=3-WRYO4AAAAJ&cstart=100&pagesize=100&sortby=pubdate&citation_for_view=3-WRYO4AAAAJ:YsrPvlHIBpEC) Guarino, N. (1995). Formal ontology, conceptual analysis and knowledge representation. International journal of human-computer studies, 43(5-6), 625-640.
[[3]](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=3-WRYO4AAAAJ&cstart=100&pagesize=100&sortby=pubdate&citation_for_view=3-WRYO4AAAAJ:DGpvO1n63MYC) Guarino, N. (Ed.). (1998). Formal ontology in information systems: Proceedings of the first international conference (FOIS'98), June 6-8, Trento, Italy (Vol. 46). IOS press.

pedropaulofb commented:
@cthoyt , below I am presenting the examples you requested, to better illustrate how the proposed mapping terms can be applied in practice. For clarity, the examples use the generic forms *Ontological Analysis* and *Ontological Unpacking*, but in real applications a mapper would typically use the more specific terms, such as *UFO-Based Ontological Analysis* and *UFO-Based Ontological Unpacking*.

### Example: Ontological Analysis

The figure below illustrates the use of **ontological analysis** to engineer a well-founded ontology and then map it to an existing one.

<img width="3294" height="1934" alt="Ontological Analysis" src="https://github.com/user-attachments/assets/1ae7819b-0397-445b-ae91-01bcb881a7a0" />

**Basic elements.**
- Each circle represents an OWL class; arrows represent relations between classes.
- The top part of Step 1 represents the domain being modeled.
- The bottom part of Step 1 represents the new ontology, engineered by the modeler and grounded in UFO via gUFO.
- Step 2 represents the mappings from the new ontology (bottom box) to an existing gUFO-grounded reference ontology (top box), shown with `skos:exactMatch`.

**Context.**
Starting from the domain, the modeler creates a new ontology with concepts `my:Person` and `my:Man`. Through ontological analysis, she assigns ontological categories:
- `my:Person` is a *(Functional Complex) Kind* (rigid sortal that provides identity).
- `my:Man` is a *(Functional Complex) SubKind* of `my:Person` (rigid specialization).

**Process.**
1. Capture domain concepts through ontological analysis by identifying their UFO categories (e.g., *Kind*, *SubKind*, *Functional Complex*) and encoding them in the new ontology.  
2. Perform ontological analysis for mapping, aligning the new ontology with an existing gUFO-grounded ontology. For example:  
   `my:Person skos:exactMatch ref:Human` (both *Functional Complex Kinds*),  
   `my:Man skos:exactMatch ref:MaleHuman` (both *Functional Complex SubKinds* of Human).

**Result.**
The ontological analysis yields a **well-founded** ontology whose classes carry explicit, theory-grounded meaning. Because the new ontology and the existing one share the same ontological categories, they can be **reliably mapped** even when labels differ (e.g., *Person* vs *Human*), improving semantic clarity and interoperability.

### Example: Ontological Unpacking

The figure below illustrates the use of **ontological unpacking** (a specific case of ontological analysis) in a mapping task.

<img width="3421" height="1868" alt="Ontological Unpacking" src="https://github.com/user-attachments/assets/7b4f0594-8315-42eb-9905-e71d65f490d7" />

**Basic elements.**
- Each circle represents an OWL class. Arrows represent relations between classes.
- The bottom box in Step 1 represents the ontology to be mapped.
- The top box in Step 1 represents a third-party well-founded reference ontology (orange classes) grounded in UFO, implemented through gUFO (blue classes).
- Step 2 represents the mappings from the ontology to be mapped (bottom box) to the reference ontology (top box), shown with `skos:exactMatch`.

**Context.**
Suppose a modeler has created an ontology that includes the classes `my:Person` and `my:Man`. She wants to map these concepts to a UFO-based reference ontology, where `ref:Human` is defined as a *(Functional Complex) Kind* (a rigid sortal that supplies identity) and `ref:HumanMale` is defined as a *(Functional Complex) SubKind* (a rigid specialization of `Human`).  

**Process.**
To perform the mapping, the modeler must first analyze the real meaning of the classes in her ontology. For example, she needs to ask whether `my:Man` is a rigid concept (something that an individual must always instantiate as long as it exists) or an anti-rigid one (something that an individual may cease to instantiate while still continuing to exist). In practical terms: can a `Person` who is once classified as `Man` cease to be a `Man` and still remain a `Person`?

By making these ontological commitments explicit, the modeler is carrying out **ontological unpacking** of her ontology. In this example, she concludes that `my:Man` is rigid, and therefore maps it to `ref:HumanMale`, a *SubKind* of `Human`. This mapping is then represented with `skos:exactMatch` links.

**Result.**
After the mappings are established, the semantics of the original ontology become clearer: the concepts `my:Person` and `my:Man` are not just descriptive categories, but carry explicit ontological meaning aligned with UFO (their meta-properties of sortality, rigidity, relational dependency, etc., are well defined and explicit).

matentzn commented:
Thank you @pedropaulofb for your interest and PR, and @cthoyt for your comments.

Before we look at the details of the PR, we need to get two important things cleared up:

--------

1. From your description in the PR (as @cthoyt points out / wonders), your intention is to use the concept you are adding here as part of the `mapping_justification` field in your SSSOM metadata. 
So in your example, you would like to express that the mapping `my:Person---[skos:exactMatch]--->ref:Human` has been determined through what you call "Ontological Analysis". Am I seeing that correctly?

---------

2. Also, it seems to me that ontological analysis (OA) is performed by a human agent. Of course, you might build an AI agent that mimics the humans' activities when performing OA, but the process seems fundamentally driven by the humans ability to map the ontological entities (eg, Classes) to the domain model (or rather, their mental model of the domain). So it seems to me that ontological analysis is fundamentally a subclass of "semapv:ManualMappingCuration". In fact, OA seems to me to be _exactly_ what we mean by `semapv:ManualMappingCuration` in the context of ontology alignment! What do you think?

----------

Before moving forward, lets answer these two fundamental questions first, then I will give more concrete feedback on your PR.

pedropaulofb committed:
feat(terms): add ManualMappingCuration as parent of OntologicalAnalysis

- Updated `semapv:OntologicalAnalysis` to declare `semapv:ManualMappingCuration` as its parent
- Applied same adjustment to all foundational variants and unpacking terms
- Cleaned up trailing whitespace in several entries

This change aligns with PR feedback confirming that Ontological Analysis
is an (mostly human) agent-driven process and should be modeled as a subclass of
`semapv:ManualMappingCuration`.

pedropaulofb commented:
@matentzn , thank you for taking the time to review the proposal.

**Regarding your first point:** yes, you’ve understood correctly. My intention is indeed to support the use of `semapv:OntologicalAnalysis` (and its foundational variants) as a value in the `mapping_justification` field of SSSOM mappings.

**As for your second point:** Given that `semapv:ManualMappingCuration` is defined as a process conducted by a (usually human) agent to determine a mapping by virtue of domain expertise, I agree that `semapv:OntologicalAnalysis` should be modeled as a subclass of `semapv:ManualMappingCuration`. In fact, when mappings are developed using foundational ontologies, that is precisely the form of manual mapping curation we are applying.

I have already updated the PR so that `semapv:OntologicalAnalysis` is declared as a specialization of `semapv:ManualMappingCuration`. Thanks again for this discussion. I’m open to any further questions you may have.

cthoyt commented:
I'm still missing a few important things that I think are necessary before making any other comments. It would be very helpful for you to do the following:

1. Explain step by step how this kind of curation works. It's not clear to me what actually happens, can you be concrete and explicit? Give a real, fully fleshed out scenario that is not a toy example.

2. Is there a real workflow that implements what you described above that we can see in action? Can you please show how this might work? Or, am I just confused and the jargon `ontological analysis` and `ontological unpacking` just mean "a person curated the mappings manually after looking at their names, descriptions, relationships, etc."

3. Give an explicit, real (i.e., not a toy example) SSSOM file where you would use this.

   If it's not clear what SSSOM files look like or how to format them, here's an example from the [Biomappings](https://github.com/biopragmatics/biomappings) project ([source](https://github.com/biopragmatics/biomappings/blob/master/src/biomappings/resources/positive.sssom.tsv)):

```tsv
subject_id	subject_label	predicate_id	object_id	object_label	mapping_justification	author_id	mapping_tool	predicate_modifier
agrovoc:0619dd9e	chain harrows	skos:exactMatch	agro:00000137	chain harrow	semapv:ManualMappingCuration	orcid:0000-0002-2627-0696		
agrovoc:0a9fbc47	scythes	skos:exactMatch	agro:00000456	scythe	semapv:ManualMappingCuration	orcid:0000-0002-2627-0696		
agrovoc:1012	border irrigation	skos:exactMatch	agro:00000066	border irrigation process	semapv:ManualMappingCuration	orcid:0000-0002-2627-0696		
agrovoc:10457	draught animals	skos:exactMatch	agro:00000116	draft animal	semapv:ManualMappingCuration	orcid:0000-0002-2627-0696		
agrovoc:1101	broadcasting	skos:exactMatch	agro:00000040	broadcast application method	semapv:ManualMappingCuration	orcid:0000-0002-2627-0696		
```

   Again, please use a real, concrete scenario that isn't a toy example to illustrate this.

4. Explain what value annotating this instead of just saying it's manually curated would add. For example, in Biomappings, we typically start with predictions, then manually curate them. It's enough for 99.9% of downstream applications to say "hey, someone manually decided this was good enough" rather than keep track of the way it was produced. In other words, the human approval is much more valuable than any other kind of mapping justification.

   How would you be using these kinds of annotations downstream?

   Alternatively, could these kinds of terms be part of the [curation_rule](https://mapping-commons.github.io/sssom/curation_rule/) field? This field lets you add more information about how curation is done.

matentzn commented:
You bring up a good point @cthoyt on `curation_rule` vs `mapping_justification`.. Its a bit of a fuzzy distinction because it `curation_rule` can indeed be interpreted as an extension of `semapv:ManualMappingCuration`.

Let's look perhaps at a concrete example where it becomes clear, for example, that `DOLCEBasedOntologicalAnalysis` and `BFOBasedOntologicalAnalysis` would come to a different conclusion when analysing a specific mapping.

Once we understand the process of `DOLCEBasedOntologicalAnalysis` better, maybe we can figure out if this is better captured as `curation_rule`.

pedropaulofb commented:
@cthoyt and @matentzn, I appreciate you reviewing the PR, and the time you invested interacting with me. I'm sorry this answer is so long, but I wanted to make sure I cleared up any doubts you have about the proposal.

> Explain step by step how this kind of curation works. It's not clear to me what actually happens, can you be concrete and explicit? Give a real, fully fleshed out scenario that is not a toy example.

Ontological analysis, always grounded in foundational ontologies, is not just a theoretical exercise used in "toy examples." I am actively using this approach in the industry, and it is widely supported in academic research. For example, the paper **"UFO: Unified Foundational Ontology"** by Guizzardi et al. (2022) provides a clear demonstration of how UFO is used for such analyses. Here's an excerpt from the paper, which shows use of ontological analysis in diverse domains:

    > For instance, it [UFO] has been used to provide conceptual clarification in domains such as:, agriculture (Neves and Cruvinel, 2021; Niederkofler et al., 2019), accounting (Blums and Weigand, 2021; Fischer-Pauzenberger and Schwaiger, 2017; Fraller, 2019; Laurier et al., 2018), business processes (Guizzardi and Wagner, 2011a; Guizzardi et al., 2016; Santos Júnior et al., 2010; van Wingerde and Weigand, 2020), biodiversity (Albuquerque et al., 2015), bioinformatics (Gonçalves et al., 2011; Rodrigues et al., 2017), branding (Elikan, 2019), communities (Almeida and Guizzardi, 2013), capabilities (Azevedo et al., 2015), competition (Sales et al., 2018c), data processing (Moura et al., 2021), decision making (Guizzardi et al., 2020; Richetti et al., 2019), design science research (Weigand et al., 2021), digital platforms (Derave et al., 2021), discrete event simulation (Guizzardi and Wagner, 2010, 2011b, 2012, 2013), economic exchanges (Porello et al., 2020), emergency and disaster management (Ferreira et al., 2015; Khantong and Ahmad, 2019; Moreira et al., 2015), engineering (Pereira et al., 2019), e-government (Barcelos et al., 2013; Detoni et al., 2018; Kˇ remen and Necaský, 2019; Pereira et al., 2020), game theory (Amaral et al., 2020c), game design (Franco et al., 2018; Kritz, 2020), goals and motivation (Azevedo et al., 2011; Guizzardi et al., 2012, 2013b), geology (Abel et al., 2015; Rodrigues et al., 2019; Vieira et al., 2020), legal issues (El Ghosh and Abdulrab, 2020, 2021; Griffo et al., 2015; Mário de Oliveira Rodrigues et al., 2020; Santos et al., 2018), money (Amaral et al., 2020b), mulsemedia (Saleme et al., 2019), organizational structures (Santos Júnior et al., 2013), programming languages (Aguiar et al., 2019), security and safety (Debbech et al., 2020; Faria et al., 2019; Kostov et al., 2017; Sales et al., 2018b; Zhou et al., 2017), services (Nardi et al., 2015), simulation for land covering and use (Grueau, 2013), smart contracts (Sharifi et al., 2020), software engineering (Duarte et al., 2018; Henderson-Sellers, 2012; Henderson-Sellers et al., 2014; Kirk and MacDonell, 2016; Maretto and Barcellos, 2014; Morales-Ramirez et al., 2015; Ruy et al., 2016; Shekhovtsov and Mayr, 2014; Sydorov et al., 2019; Wang et al., 2014b), software requirements (Bernabé et al., 2019; Duarte et al., 2018, 2021; Guizzardi et al., 2014; Li et al., 2015; Negri et al., 2017), telecommunication networks (Barcelos et al., 2009; Rita de Cássia et al., 2012), treatment (Johannesson and Perjons, 2020), tourism (AlpineBits Alliance, 2021), trust (Amaral et al., 2019), value (Gailly et al., 2016; Guarino et al., 2016; Sales et al., 2017), waste management (Ahmad et al., 2018).
    > Moreover, UFO and ontologies built with it have been used to analyze, reengineer, or integrate many modeling languages and standards in different domains:, ArchiMate (Almeida et al., 2009; Amaral et al., 2020a; Azevedo et al., 2011, 2015; Griffo et al., 2017; Sales et al., 2018a, 2019), ARIS (Santos Júnior et al., 2010, 2013), DEMO (Poletaeva et al., 2017), ISO/IEC 24744 (Ruy et al., 2014), ITU-T G.805 (Barcelos et al., 2011), BPMN (Guizzardi and Wagner, 2011a), RM-ODP (Almeida and Guizzardi, 2013), TOGAF (Almeida et al., 2009), Tropos and i* (Guizzardi et al., 2013b,c; Franch et al., 2011), UML (Costal et al., 2011; Guizzardi, 2005).

For further examples using other foundational ontologies, I refer you to the special issue of the [journal *Applied Ontology* (Volume 17, Issue 1)](https://journals.sagepub.com/toc/apo/17/1), which features seven of the most important foundational ontologies used for well-founded analysis, modeling, and mapping.

For their practical use, some foundational ontologies are directly implemented in semantic web formats. For instance, **BFO**, **GFO**, and **PROTON** have direct OWL implementations. Additionally, some have versions coded in these languages, such as **DOLCE** and **UFO**, (see [Trojahn et al. (2022)](https://journals.sagepub.com/doi/pdf/10.3233/SW-210447)). This allows for their direct use in machine-readable applications, further enabling the kind of interoperability we are building at Health-RI.

As a concrete example, [consider my recent work with Health-RI](https://w3id.org/health-ri/semantic-interoperability), a Dutch foundation. Our goal is to enable semantic interoperability across various Dutch and European healthcare organizations. In this initiative, we aim to offer a domain reference ontology. Our idea is that concepts from other domain models can be mapped to this ontology, making explicit their semantics (i.e., we want our ontology to always be the `object` in our curated SSSOM mappings).

> Is there a real workflow that implements what you described above that we can see in action?

Yes, there is a real workflow being developed based on what I described. As the initiative is still new, I am currently working on this specification. The partial description [can already be seen in here](https://health-ri.github.io/semantic-interoperability/method/mapping-set/). We will have a team specialized in performing the mappings between the concepts from existing healthcare standards and our reference ontology.

> Or, am I just confused and the jargon ontological analysis and ontological unpacking just mean "a person curated the mappings manually after looking at their names, descriptions, relationships, etc."

Ontological analysis, as I mentioned in my last message, can indeed be classified under "manual curation." However, it is not just any manual curation: it is done systematically and relies on formalized logical axioms from foundational ontologies, trying to identify what are the meta-properties of the concepts to be aligned.

So yes, in the ontological analysis, a person defines the mappings manually after carefully analyzing names, descriptions, relationships, and other contextual elements. But the key point is that this process is meant to capture the *ontological commitment* of the model. It involves a kind of "re-engineering" to uncover and clarify what is *really* being represented.

I wouldn't simplify this to just "a person curated the mappings manually after looking at their names, descriptions, relationships, etc." To me, that would be like saying all algorithmic methods in SEMAPV are simply "algorithmic method," since they're all just "algorithms performed by a computational agent." The label hides the nature, complexity, and rationale of the method being used.

> "Give an explicit, real (i.e., not a toy example) SSSOM file where you would use this."

In [the page I mentioned before](https://health-ri.github.io/semantic-interoperability/method/mapping-set/), the `mapping_justification` column should ideally use the proposed `semapv:OntologicalAnalysis` or `semapv:OntologicalUnpacking` to more precisely describe the justification behind the mappings.

Note, however, that for now the value used is just `semapv:ManualMappingCuration`, as a more specific term is not yet available in the vocabulary. The introduction of these more fine-grained justifications is exactly what this PR aims to address.

> Explain what value annotating this instead of just saying it's manually curated would add. (...) How would you be using these kinds of annotations downstream?

The key added value here is **reliability**, but not in the same sense captured by the `confidence` slot in SSSOM. What we're doing is not just asserting a loose correspondence between two terms from semantic web ontologies. Instead, we are identifying the semantics (in a more precise level) of a concept defined in a knowledge artifact.

In these cases, one artifact (that will have concepts mapped as `object`) will be a higher-level reference ontology, while the other (concepts as `subject`) will be lower-level representation like a data model or application-specific vocabulary. I've tried to elaborate on this idea further [in this page](https://health-ri.github.io/semantic-interoperability/method/mapping-strategy/).

So, a concept that explicitly indicates that someone performed an ontological analysis to define a mapping significantly increases the reliability of that mapping. A user of it would know that the provider relied on the axioms of a foundational ontology to systematically evaluated the concept being mapped.

> Alternatively, could these kinds of terms be part of the curation_rule field?

Finally, although it is technically possible to include this information in the `curation_rule` field, doing so would significantly hinder many of the intended benefits.

This type of annotation allows downstream consumers (e.g., data integrators, knowledge graph builders, semantic validators) to understand the rationale behind the mapping, prioritize mappings for reuse, and filter them accordingly.

Having it as a value in the `mapping_justification` field instead makes the mappings more FAIR. They would become easier to find, interoperate with, and reuse. For example, a user could query specifically for mappings derived via ontological unpacking based on a particular foundational ontology, knowing that the mapping was evaluated with the same ontological commitments they are working with.

By contrast, embedding this information in `curation_rule` would make such mappings harder to query, less interoperable, and more error-prone, as it would require parsing free text or complex procedure descriptions rather than relying on a well-defined controlled vocabulary.

pedropaulofb commented:
@cthoyt and @matentzn, apologies for the long previous message. Please let me know if you have any feedback or considerations about it, or if I should simplify it further to help us move the PR forward. Thanks!

matentzn commented:
@pedropaulofb just so you know, I have not forgotten about your request - I just hit a few deadlines at once, but it remains firmly on my todo.

> By contrast, embedding this information in curation_rule would make such mappings harder to query, less interoperable, and more error-prone, as it would require parsing free text or complex procedure descriptions rather than relying on a well-defined controlled vocabulary.

This is not quite true, I think you are thinking of `curation_rule_text`. `curation_rule` refers to a resource identifier just like the mapping justifications do.

pedropaulofb commented:
> @pedropaulofb just so you know, I have not forgotten about your request - I just hit a few deadlines at once, but it remains firmly on my todo.
> 
> > By contrast, embedding this information in curation_rule would make such mappings harder to query, less interoperable, and more error-prone, as it would require parsing free text or complex procedure descriptions rather than relying on a well-defined controlled vocabulary.
> 
> This is not quite true, I think you are thinking of `curation_rule_text`. `curation_rule` refers to a resource identifier just like the mapping justifications do.

Thanks for the follow-up, @matentzn. I completely understand about the deadlines. I'm often in the same situation myself.

You are absolutely right: I mistakenly confused `curation_rule` with `curation_rule_text`. Thanks for the clarification. However, I believe the other points still hold.

Would you prefer that I revise this part now, or would it be better to wait until you have reviewed everything more thoroughly?

cthoyt commented:
I'm feeling very overwhelmed by your response. It's both long and I don't think actually addressed my key concerns. Let me try and rephrase what I am looking for:

1. One or both of the following: 
   1. a step-by-step tutorial that teaches a curator how to curate doing the analysis. Imagine that I am the target audience, who doesn't know any of the terminology you use ahead of time. Consider a learner likely has limited bandwidth for following links to prior material.
   2. a link to an open source, documented programmatic workflow that I can run myself that implements/supports the curation that you're describing
2. An example SSSOM file that has a real scenario where you did this, demonstrating where you would use these terms

Please let me know if the expectations are unclear. I am happy to continue discussion iff you can address these.

pedropaulofb commented:
Dear @cthoyt, my intent is not to define or standardize a community workflow, but to provide terms so curators can register how a mapping was derived. To make this operational without being prescriptive, here is a short, non-normative checklist for UFO-based ontological unpacking.

```
1. Select the foundational ontology (here: UFO) and the scope of concepts to map.
2. Elicit ontological commitments ("truthmakers") for each concept in the source/target by clarifying what kinds of entities and relations must exist if the concept description were true.
3. Assess meta-properties (e.g., rigidity, identity, dependence) of each concept.
4. Choose the foundational ontology's consistent pattern that best captures the concept (e.g.,in UFO, Role *vs.* Kind).
5. Choose the mapping predicate (e.g., `skos:exactMatch`, `skos:broadMatch`) consistent with the alignment.
6. Record in SSSOM: set `mapping_justification = semapv:UFOBasedOntologicalUnpacking` (proposed).
```

As for a concrete SSSOM file that uses this mapping justification, I don’t have one for now, but it will look like the snippet below. Concrete cases will be available soon as our initiative progresses and artifacts can be published.

```
subject_id	subject_label	predicate_id	object_id	object_label	mapping_justification	author_id	mapping_date
hc:Patient	Patient	hriv:isExactMeaningOf	hrio:HealthcarePatient	Healthcare Patient	semapv:UFOBasedOntologicalAnalysis	https://orcid.org/0000-0000-0000-0000	2025-09-01
vet:Patient	Patient	hriv:isExactMeaningOf	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
ghc:Patient	Patient	hriv:hasBroaderMeaningThan	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
pvet:Patient	Patient	hriv:hasBroaderMeaningThan	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
pvet:PetAnimal	Pet Animal	hriv:hasNarrowerMeaningThan	hrio:NonHumanAnimal	Non-Human Animal	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
ghc:Animal	Animal	hriv:isExactMeaningOf	hrio:Animal	Animal	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
```

Kindly inform me whether these provided examples are sufficient or if further additions are necessary.

cthoyt commented:
Can you please qualify this SSSOM file with a prefix map? Then, I can understand the vocabularies that these entities come from.

I don't think this tutorial is sufficient for me to understand to be able to understand what these steps are. Can you please use the examples from the SSSOM you copied to illustrate each step?

pedropaulofb commented:
Thanks, @cthoyt. I have qualified the example with a prefix map in the SSSOM header. I have also expanded the checklist and annotated each step with references to the example rows below.

```
Non-normative checklist (UFO-based ontological unpacking), with examples

1. **Select the foundational ontology and scope.**
   *Here: UFO; scope = patient/person/animal concepts across four sources.*

2. **Elicit ontological commitments (“truthmakers”).**
   *Example:* `hrio:HealthcarePatient` ≈ a **human** under care within a healthcare context; `hrio:VeterinaryPatient` ≈ a **non-human animal** under care in a veterinary context.

3. **Assess meta-properties & UFO category.**
   *Example:* `Patient` behaves as a **Role** (anti-rigid; depends on a care relation), while `Animal`/`NonHumanAnimal` are **(sub)Kinds** (rigid, relational independent).

4. **Align patterns per UFO.**
   *Examples:*
   • `hc:Patient` aligns to a **Role on Human** -> matches `hrio:HealthcarePatient` (human patient).
   • `vet:Patient` aligns to a **Role on NonHumanAnimal** -> matches `hrio:VeterinaryPatient`.
   • `pvet:PetAnimal` aligns to a **SubKind** more specific than `hrio:NonHumanAnimal`.

5. **Choose the mapping predicate consistent with the alignment.**
   *Examples:*
   • **Exact** meaning within the corresponding context: rows 1–2 use `hriv:isExactMeaningOf`.
   • **Broader/Narrower**: rows 3–4 use `hriv:hasBroaderMeaningThan`; row 5 uses `hriv:hasNarrowerMeaningThan`.

Here, instead of SKOS, I am using a own-defined vocabulary to map the concepts' meanings.

6. **Record provenance in SSSOM.**
   In this illustration I use the proposed justification `semapv:UFOBasedOntologicalUnpacking`. 

```

SSSOM example (with prefix map):

```
# sssom_version: 1.1
# mapping_set_id: https://example.org/ufo-unpacking-demo
# mapping_set_description: Illustrative slice showing UFO-based unpacking on patient/person/animal.
# curie_map:
#   hc: http://example.org/hc/
#   vet: http://example.org/vet/
#   ghc: http://example.org/ghc/
#   pvet: http://example.org/pvet/
#   hrio: https://w3id.org/health-ri/ontology#
#   hriv: https://w3id.org/health-ri/mapping-vocabulary#
#   skos: http://www.w3.org/2004/02/skos/core#
#   semapv: https://w3id.org/semapv/vocab/
subject_id	subject_label	predicate_id	object_id	object_label	mapping_justification	author_id	mapping_date
hc:Patient	Patient	hriv:isExactMeaningOf	hrio:HealthcarePatient	Healthcare Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
vet:Patient	Patient	hriv:isExactMeaningOf	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
ghc:Patient	Patient	hriv:hasBroaderMeaningThan	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
pvet:Patient	Patient	hriv:hasBroaderMeaningThan	hrio:VeterinaryPatient	Veterinary Patient	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
pvet:PetAnimal	Pet Animal	hriv:hasNarrowerMeaningThan	hrio:NonHumanAnimal	Non-Human Animal	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
ghc:Animal	Animal	hriv:isExactMeaningOf	hrio:Animal	Animal	semapv:UFOBasedOntologicalUnpacking	https://orcid.org/0000-0000-0000-0000	2025-09-01
```

cthoyt commented:
@pedropaulofb I think there might be a misunderstanding, I want to see usage of real vocabularies, not made up ones from example.org. I can't consider an example that doesn't use real vocabularies. I will judge this by whether I can expand the CURIE to a URL that actually resolves and says something by the entity

can you also please use markdown formatting so I don't have to read from the preformatted text block? Maybe it's best to move this guide into a google doc that I can comment/suggest on, because I have a feeling it will take several iterations to make it into something that any curator can understand/follow

pedropaulofb commented:
Dear @cthoyt and @matentzn,

Unfortunately, it looks like we are not making progress here. I still do not have concrete, real-world examples (as I mentioned in earlier comments), and the academic references were not sufficient to justify the proposed additions at this time. Given that, I am closing this PR.

Thank you for your time reviewing this, and congratulations again on the great work you are doing here.

cthoyt commented:
@pedropaulofb I agree, this is probably the right outcome for now. When your project is a bit more mature and can give a better demonstration, then we can resume the discussion.

For now, I would suggest using `semapv:ManualMappingCuration` for `mapping_justification` and using free text in `curation_rule_text` (or using your own vocabulary for `curation_rule`) to capture what you're trying to do

matentzn commented:
I know this is closed, but I want to push back on two arguments you gave @pedropaulofb for using curation_rule. In fact, I think this is _exactly_ what was intended for the `curation_rule` field:

> Having it as a value in the mapping_justification field instead makes the mappings more FAIR. 

How so?

Example 1:

| subject_id | ... | mapping_justification | curation_rule |
| ---------- | ---|---------------------- | ------------- |
| s:1 | ... |  semapv:UFOBasedOntologicalUnpacking | |

seems to me equally FAIR and expressive as

Example 2:

| subject_id | ... | mapping_justification | curation_rule |
| ---------- | ---|---------------------- | ------------- |
| s:1 | ... |  semapv:ManualMappingCuration | myrule:UFOBasedOntologicalUnpacking |

In both cases we use Resource identifiers. Can you explain?

> They would become easier to find, interoperate with, and reuse. For example, a user could query specifically for mappings derived via ontological unpacking based on a particular foundational ontology, knowing that the mapping was evaluated with the same ontological commitments they are working with.

This is easily possible with example 2 as well. In fact, sssom-py comes with functionality do query/filter such sets!

> By contrast, embedding this information in curation_rule would make such mappings harder to query, less interoperable, and more error-prone, as it would require parsing free text or complex procedure descriptions rather than relying on a well-defined controlled vocabulary.

We have already discussed this is not the case.

Apart from Charlies concerns (he is not from the hardcore ontology world - I do understand what `UFOBasedOntologicalUnpacking` is so did not personally need an explanation), I would like to really clarify why this is not better modelled as curation rules. I would even consider adding these terms, but typing them as curation rules rather then justifications (as a reminder, we want that to keep the list of justifications smaller and more easy to use).

pedropaulofb commented:
Dear @matentzn, thanks for looking into this even though the PR was closed. I have re-opened it so we can keep the discussion and explore a potential contribution.

> In both cases we use Resource identifiers. Can you explain?

You are right. modeling this via `curation_rule` would work, and FAIRness would not be compromised. It would still be clear that a more specific, ontology-grounded approach was used than generic manual curation.

> I would like to really clarify why this is not better modelled as curation rules. I would even consider adding these terms, but typing them as curation rules rather then justifications (as a reminder, we want that to keep the list of justifications smaller and more easy to use).

Understood. Just to confirm my understanding: are `curation_rule` values intended to be terms provided by the vocabulary, or user-defined project terms? If the latter, I can close the PR and encode these as `curation_rule` instances in my mapping set. If the former, I can adjust the PR to propose them as rule terms rather than justifications.

Please let me know which path you would prefer.

matentzn commented:
> Understood. Just to confirm my understanding: are curation_rule values intended to be terms provided by the vocabulary, or user-defined project terms? If the latter, I can close the PR and encode these as curation_rule instances in my mapping set. If the former, I can adjust the PR to propose them as rule terms rather than justifications.

The curation_rule values are _typically_ defined as controlled vocabulary terms locally, by the entity that is creating the mapping set (usually on organisation level). I would propose this:

1. Produce a valid SSSOM files with the curation_rule terms.
2. Make sure the curation_rule terms can resolve, say using w3id.org or any other URL based system
3. Share it here when you are done

I will inspect it when its done, and then we can have a discussion of a centralised place for curating _generally useful curation rules_. 

BTW, we checked some of your SSSOM files and there are some validation issues. Did you every try to validate, like 

- https://sss-om-validate.streamlit.app/?
- https://incenp.org/dvlpt/sssom-java/apidocs/org/incenp/obofoundry/sssom/Validator.html
- https://gbv.github.io/sssom-js/

pedropaulofb commented:
Thank you very much for the info, @matentzn.

> I will inspect it when its done, and then we can have a discussion of a centralised place for curating generally useful curation rules.

Given that the PR content is no longer relevant, I believe it would be better to close this PR and open a new issue that links to it after I have produced the required information. Are you okay with that?

> BTW, we checked some of your SSSOM files and there are some validation issues. Did you every try to validate, like
> 
> https://sss-om-validate.streamlit.app/?
> https://incenp.org/dvlpt/sssom-java/apidocs/org/incenp/obofoundry/sssom/Validator.html
> https://gbv.github.io/sssom-js/

I appreciate your input. I tried to run local tests with sssom-py, but it does not appear to fully align with the latest sssom version (1.1), correct? I have not tried the two online services that you sent me before. After receiving them from you, I could find and fix some small issues, but unfortunately, they are giving me results that do not fully correspond to the content being tested.

Since I was unable to use sssom-py to convert to ttl, I wrote a basic converter [to produce my own ttl file](https://w3id.org/health-ri/semantic-interoperability/mappings). Are you aware of a service that can check the validity of sssom files serialized in Turtle?

Thank you in advance for your guidance.

matentzn commented:
If you need SSSOM 1.1 you will have to use

https://incenp.org/dvlpt/sssom-java/sssom-cli/index.html#a1.4._Input_validation

But afaik this only works for TSV. So I would validate the source TSV (assuming there is one) and using your rdf conversion script until we finish the spec for the RDF conversion.

Yes, if you like to move the discussion back to the/a issue that would be great!

pedropaulofb commented:
Dear @matentzn,

> If you need SSSOM 1.1 you will have to use
> https://incenp.org/dvlpt/sssom-java/sssom-cli/index.html#a1.4._Input_validation

Thank you for the tip! I will check that.

> But afaik this only works for TSV. So I would validate the source TSV (assuming there is one) and using your rdf conversion script until we finish the spec for the RDF conversion.

We do have a tsv source file: 

- https://w3id.org/health-ri/semantic-interoperability/mappings
- https://github.com/Health-RI/semantic-interoperability/tree/main/mappings

> Yes, if you like to move the discussion back to the/a issue that would be great!

I am closing this PR, as agreed and will open a related issue in the future. 

Kind regards.

matentzn commented:
Thank you @pedropaulofb!
